### PR TITLE
[JEP-234] Customizable header proposal

### PR DESCRIPTION
**Pull request by imonteroperez**
### Highlights
- Proposal on having a customizable header (not only UI capabilities) it is also providing extra functionality and better integration
- See: https://github.com/jenkinsci/jenkins/pull/5909

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes

### CODE REVIEW

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Thanks @imonteroperez ! I am not sure why this particular change requires a JEP. I am not against it. I am requesting changes because of the incorrectly pre-assigned JEP number. With the current numeration strategy it should be rather JEP-234

File: jep/401/README.adoc
Comment (imonteroperez) [reply jep/401/README.adoc]: Updated in https://github.com/jenkinsci/jep/pull/380/commits/781a489560666583e9623950112ab30a9010eb2d

**Comment by imonteroperez** [comment]
> Thanks @imonteroperez ! I am not sure why this particular change requires a JEP. I am not against it. I am requesting changes because of the incorrectly pre-assigned JEP number. With the current numeration strategy it should be rather JEP-234

Thanks for heads up on numeration! Updated in https://github.com/jenkinsci/jep/pull/380/commits/781a489560666583e9623950112ab30a9010eb2d

**Comment by batmat** [comment]
A note on the process: as mentioned in JEP-1 linked from the README, PR normally don't set JEP numbers. This number gets chosen when the PR gets merged.

> When they are done, the Draft JEP will have an official JEP number and the submission PR will have been merged to a matching folder (for example, jep/1) in the master branch.

Given Oleg's answer, I assume we can keep 234, but let's remember for next time :)

File: jep/234/README.adoc
Comment (batmat) [inline jep/234/README.adoc:138]: This kind of block should use `[source,java]` (see the [doc](https://docs.asciidoctor.org/asciidoc/latest/verbatim/source-blocks/)). I'm not even sure \`\`\` is actually asciidoc.

File: jep/234/README.adoc
Comment (batmat) [inline jep/234/README.adoc:199]: "i.e." or "e.g." you mean?

**Review by batmat — APPROVED** [review]
Overall LGTM. Many wordsmithing suggestions and a few bigger changes proposed.

But again, nothing blocking I could see on the design side of the subject. (hence my approval)

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc]: Updated in https://github.com/jenkinsci/jep/pull/380/commits/080272a6cf84a36a56e13aa21fbe9bd3aab3cb9d

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc]: Updated in https://github.com/jenkinsci/jep/pull/380/commits/6dcf673f99ca6e79460396d916d6725448365f06

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc]: Updated in https://github.com/jenkinsci/jep/pull/380/commits/0d9f3aa1a39e53cbedf50929f336176fc5dcf88c

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc:106]: JEP proposes to limit the number of active headers to one (so this is counting the default one too)

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc:112]: I saw that section in some JEPs, but if it sounds redundant/repetitive no issues on removing it.

Comment (imonteroperez) [reply jep/234/README.adoc]: Removed on: https://github.com/jenkinsci/jep/pull/380/commits/ba6d169b0e16c46ec3d4f426a4e5387e002b8947

**Review by amuniz — APPROVED** [review]
Looks good.

**Comment by jglick** [comment]
> I assume we can keep 234

Better to use `0000` as directed by JEP-1 and assumed by #376.

File: jep/234/README.adoc
Comment (jglick) [inline jep/234/README.adoc:1]:
```suggestion
= JEP-0000: Customizable Jenkins header
```

plus file rename

File: jep/234/README.adoc
Comment (jglick) [inline jep/234/README.adoc:16]:
```suggestion
| 0000
```

File: jep/234/README.adoc
Comment (jglick) [inline jep/234/README.adoc:180]: Leave this sort of thing to the core PR please. Duplicated here and just liable to drift out of date with reality, and also to attract review comments that belong on the actual code.

File: jep/234/README.adoc
Comment (jglick) [inline jep/234/README.adoc:237]: Last I checked `invokeStatic` does not work on plugin classes. Did you actually test this? (In `RealJenkinsRule` or using `java -jar jenkins.war`, _not_ using `JenkinsRule` or `mvn hpi:run`!)

I think you rather want to ensure that `it` is correctly bound when rendering Jelly.

File: jep/234/README.adoc
Comment (jglick) [inline jep/234/README.adoc:238]:
```suggestion
<span class="hidden-xs hidden-sm">${label}—${userName}</span>
```

File: jep/234/README.adoc
Comment (jglick) [inline jep/234/README.adoc:250]: This needs to be fleshed out. The sample plugin involves considerable duplication of core code, references to core resources such as images which might disappear without notice, etc.

File: jep/234/README.adoc
Comment (jglick) [inline jep/234/README.adoc:267]: Please delete `CustomHeaderExtensionFilter`. AFAICT it accomplishes nothing and just confuses readers.

File: jep/234/README.adoc
Comment (jglick) [inline jep/234/README.adoc:135]: Does not belong here.

File: jep/234/README.adoc
Comment (jglick) [inline jep/234/README.adoc:1]: needs to be in a file named `jep/0000/README.adoc`

Comment (imonteroperez) [reply jep/234/README.adoc]: Updated in https://github.com/jenkinsci/jep/pull/380/commits/e1d733370cf1bdba8d4591b9190f55999e5e5370

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc]: `CustomHeaderExtensionFilter` removed in: https://github.com/imonteroperez/custom-header-plugin/commit/ebdae6039adbb318683a69bd923834c549f47ab1

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc]: Tested using `java -jar jenkins.war` and working without issues

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc]: Based on [Reasoning](https://github.com/jenkinsci/jep/tree/master/jep-template/0000#reasoning) description, we can explain here why particular design decisions were made and also why current approaches were discarded. So, based on an illustrative example of customization:

* I provide a list of discarded current approaches (and why are discarded)
* I provide some source code there to illustrate the proposed design. 

Added an explicit mention about final source code may differ and reference to Reference Implementation section

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc]: Ditto: https://github.com/jenkinsci/jep/pull/380#discussion_r750143104

File: jep/234/README.adoc
Comment (imonteroperez) [reply jep/234/README.adoc]: Updated in https://github.com/jenkinsci/jep/pull/380/commits/e93d9a8da01316bc45deb11c3ccad415d210c302

**Comment by imonteroperez** [comment]
@amuniz @batmat although you approved the content of the JEP proposal, some relevant changes were added regarding to backward compatibility. I would appreciate your re-review. Thanks in advance!

**Review by amuniz — APPROVED** [review]

**Comment by imonteroperez** [comment]
@oleg-nenashev @batmat @jglick would you mind re-reviewing JEP content now. Seems implementation is getting approvals in https://github.com/jenkinsci/jenkins/pull/5909 so based on the JEP procedure maybe we could merge this (keeping Draft state), WDYT?

File: jep/0000/README.adoc
Comment (jglick) [inline jep/0000/README.adoc:78]: ⇒ **Motivation** + **Reasoning**. **Abstract** should be short and to the point.

File: jep/0000/README.adoc
Comment (jglick) [inline jep/0000/README.adoc:87]: Check JEP-1 and write in the present tense.

File: jep/0000/README.adoc
Comment (jglick) [inline jep/0000/README.adoc:90]: The reverse I guess?

File: jep/0000/README.adoc
Comment (jglick) [inline jep/0000/README.adoc:103]: Again, leave details like this to the reference implementation.

File: jep/0000/README.adoc
Comment (jglick) [inline jep/0000/README.adoc:129]: You mean, a fork?

File: jep/0000/README.adoc
Comment (jglick) [inline jep/0000/README.adoc:132]: Also the critical problem with this plugin: when core resources change, you get a diff that does not apply, and it is hell to recreate.

File: jep/0000/README.adoc
Comment (jglick) [inline jep/0000/README.adoc:136]: Not sure why this section exists. Just delete from here to **Backwards Compatibility** IMO.

File: jep/0000/README.adoc
Comment (jglick) [inline jep/0000/README.adoc:255]: Discussion of the diffpatch plugin belongs in **Reasoning**. Here just describe how you test _your_ proposal.

Also BTW https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

File: jep/0000/README.adoc
Comment (jglick) [inline jep/0000/README.adoc:260]: And uses a fully duplicated `headerContent.jelly`. Are there any examples of incremental modification, if that is supported?

File: jep/0000/README.adoc
Comment (MarkEWaite) [inline jep/0000/README.adoc:116]: Optional - since we're already inside the "Reasoning" section, I don't think we need to refer readers to the Reasoning section for more details.
```suggestion
Also, other alternatives have been evaluated as workaround like the https://github.com/stephenc/diffpatch-maven-plugin[diffpatch-maven-plugin] with no satisfactory results.
```

File: jep/0000/README.adoc
Comment (MarkEWaite) [inline jep/0000/README.adoc:227]: Optional - helps with the navigation if we make the reference a hyperlink
```suggestion
* See the sample implementation provided in the <<Reference Implementation>> section.
```

File: jep/0000/README.adoc
Comment (MarkEWaite) [inline jep/0000/README.adoc:97]: Make this a hyperlink for easier navigation
```suggestion
The unique existing approach based on the https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin] has limited capabilities (see <<Reasoning>> section for more details).
```

File: jep/0000/README.adoc
Comment (MarkEWaite) [inline jep/0000/README.adoc:131]: Optional - make the reference a hyperlink for easier navigation
```suggestion
Given existing approaches do not fulfill this example, we will explore the alternative of a Jenkins user that wants to update the Jenkins header through an ad-hoc Jenkins plugin that follows the principles provided in the <<Specification>> section.
```

File: jep/0000/README.adoc
Comment (MarkEWaite) [inline jep/0000/README.adoc:133]: Optional - hyperlink for easier navigation by readers
```suggestion
The following source code is provided just for illustrative purposes on reasoning why particular design decisions were made. It may differ from the final implementation (please go to <<Reference Implementation>> section to see the final proposed changes in source code)
```

File: jep/0000/README.adoc
Comment (MarkEWaite) [inline jep/0000/README.adoc:240]: Optional - Hyperlink for navigation
```suggestion
Compatibility check for `PartialHeader` will be based on evaluating the field `compatibilityHeaderVersion`. When an incompatible change is made in the header (like the search form API), compatibility header version should be increased. See <<Reference Implementation>> for futher details.
```

**Review by MarkEWaite — APPROVED** [review]
I believe this is ready to be merged and assigned JEP-234

I made some optional comments.  No change is required by me before this is merged.  Thanks!

File: jep/0000/README.adoc
Comment (imonteroperez) [reply jep/0000/README.adoc]: Updated in https://github.com/jenkinsci/jep/pull/380/commits/cc194ad3c5ae36f5d8e64a1d615db383a60ecbe7

File: jep/0000/README.adoc
Comment (imonteroperez) [reply jep/0000/README.adoc]: Updated in https://github.com/jenkinsci/jep/pull/380/commits/cc194ad3c5ae36f5d8e64a1d615db383a60ecbe7

File: jep/0000/README.adoc
Comment (imonteroperez) [reply jep/0000/README.adoc]: Default one uses `Integer.MIN_VALUE` https://github.com/jenkinsci/jenkins/blob/e4ddcf1234bb1dccfaa529413d9b16e5d65c31e6/core/src/main/java/jenkins/views/JenkinsHeader.java#L10. BTW removed on https://github.com/jenkinsci/jep/pull/380/commits/cc194ad3c5ae36f5d8e64a1d615db383a60ecbe7 to avoid confussion

File: jep/0000/README.adoc
Comment (imonteroperez) [reply jep/0000/README.adoc]: Yep. Updated on https://github.com/jenkinsci/jep/pull/380/commits/cc194ad3c5ae36f5d8e64a1d615db383a60ecbe7

File: jep/0000/README.adoc
Comment (imonteroperez) [reply jep/0000/README.adoc]: Added on https://github.com/jenkinsci/jep/pull/380/commits/cc194ad3c5ae36f5d8e64a1d615db383a60ecbe7

File: jep/0000/README.adoc
Comment (imonteroperez) [reply jep/0000/README.adoc]: As mentioned in the document:

> The following source code is provided just for illustrative purposes on the reasoning why particular design decisions were made. It may differ from the final implementation.

File: jep/0000/README.adoc
Comment (imonteroperez) [reply jep/0000/README.adoc:87]: As mentioned, prefered to provide some code for illustrative purposes. In this case about how `pageHeader.jelly` resource is reduced from +100 lines to 30

File: jep/0000/README.adoc
Comment (imonteroperez) [reply jep/0000/README.adoc]: I will work on a test example in the reference implementation `custom-header-plugin`. Applied one-sentence-per-line here https://github.com/jenkinsci/jep/pull/380/commits/b6b353c1adca78cf8f2432a91cc6da49467906b7

Comment (imonteroperez) [reply jep/0000/README.adoc]: https://github.com/imonteroperez/custom-header-plugin/issues/1

File: jep/0000/README.adoc
Comment (imonteroperez) [reply jep/0000/README.adoc]: Updated description in https://github.com/jenkinsci/jep/pull/380/commits/d1f9b7336ae776ac8293a4bb2d8b252493d19255 to reflect what that reference implementation prototype does.

**Review by oleg-nenashev — APPROVED** [review]
I will go ahead and merge it

**Comment by batmat** [comment]
Thank you Oleg!

**Comment by batmat** [comment]
@oleg-nenashev @jenkinsci/jep-editors can we merge this PR and assign it an official number pretty please? This is ready for merge FWICT, and this is somehow blocking https://github.com/jenkinsci/jenkins/pull/5909 merge (which is ready too).

And it would be cleaner to merge first the JEP before the reference implementation, hence my request here :-).

Thanks!

**Comment by oleg-nenashev** [comment]
Yes I'm about doing so. Was unable to do it yesterday due to yet another
family emergency.

**Comment by oleg-nenashev** [comment]
Please note that there are at least four JEP editors employed by CloudBees
who could do such formalities. The companies will come out to participate
in the community governance

**Review by MarkEWaite — APPROVED** [review]

**Comment by batmat** [comment]
@MarkEWaite are you able to do the landing of this PR as a draft? Oleg agrees with the content AFAIK, but would welcome some help on the process if we can.

**Comment by MarkEWaite** [comment]
I'm happy to do the merge @oleg-nenashev if you're willing to "check my work" that I didn't make some obvious mistake.  I believe the merge steps are:

* Create a commit that assigns JEP-234 to this PR (as was done for another JEP in c691756e7c2217f27e09c6990fc76df7d1082fb6)
* Create a commit that adds JEP-234 to top level README (as was done for another JEP in 42bf67e77ce9c029b09650260289ea81c907add9)
* Merge the branch to master

**Comment by timja** [comment]
@MarkEWaite
There's a script:
https://github.com/jenkinsci/jep/tree/master/jep/1#approve-as-draft

Probably should be more discoverable than buried in that monster though

**Comment by MarkEWaite** [comment]
> There's a script: https://github.com/jenkinsci/jep/tree/master/jep/1#approve-as-draft

Thanks for that pointer!  I adjusted the results slightly (leaving the delegate status as TBD instead of using the value 'not delegated' that was set by the script.  If that was a mistake, we can correct it with another pull request.

**Comment by MarkEWaite** [comment]
This pull request was merged by following the script.  After merging it from the command line, the pull request was still open, so I closed it.  In the future, I may use the `gh pr merge` command so that the pull request will be listed by GitHub as "merged" instead of "closed"

**Comment by timja** [comment]
Looks like its only on your repo atm @MarkEWaite

**Comment by jglick** [comment]
The script may need to be adjusted depending on how you name your remotes—I use `origin` for the @jenkinsci version and `fork` for my @jglick copy, but there is no standard.

**Comment by jglick** [comment]
And yes if the script is used correctly, the final `git push` will automatically result in the PR being considered merged, because GitHub will detect that the PR branch head is now an ancestor of the base branch head.

**Comment by MarkEWaite** [comment]
I missed the `push` to upstream.  I use an inverted naming convention compared to @jglick.  Merging now with the GitHub merge pull request button

**Comment by jglick** [comment]
> Merging now with the GitHub merge pull request button

Also not what the recommended script calls for. It looks like you did some commits by hand to set the number and so on which the proposed script handles a bit differently. Should not matter now, just FYI for next time.

**Pull request merged**
